File: solution/189.rotate-array/rotate-array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function (nums, k) {
      let len = nums.length
      if (!len ||  !k || k ===len ) {
        return
      }
      function swap(i, j) {
        let temp = nums[i]
        nums[i] = nums[j]
        nums[j] = temp

      }
      for (let i = 0, j = (len + len - k - 1) % len;i < j;i++ , j--) {
        swap(i, j)
      }
      for (let i = (2 * len - k) % len, j = len - 1;i < j;j-- , i++) {
        swap(i, j)
      }
      for (let i = 0, j = len - 1;i < j;i++ , j--) {
        swap(i, j)
      }
      console.log(nums)
    };

File: solution/189.rotate-array/rotate-array.c
void swap(int i,int j,int *nums){
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
}
void rotate(int* nums, int numsSize, int k) {
  int len = numsSize,i ,j;
  if (!len || !k || k == len) {
    return;
  }
  for ( i = 0, j = (len + len - k - 1) % len; i < j; i++, j--) {
    swap(i, j,nums);
  }
  for ( i = (2 * len - k) % len, j = len - 1; i < j; j--, i++) {
    swap(i, j,nums);
  }
  for ( i = 0, j = len - 1; i < j; i++, j--) {
    swap(i, j,nums);
  }
}